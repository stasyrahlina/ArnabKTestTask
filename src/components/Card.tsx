import React, { useContext } from 'react';
import { ContextApp, ContextType } from '../context';
import { Message } from '../Api'

import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';

import { 
  CustomButton,
  CustomCardActions,
  StyledCard,
} from '../styles'

const Card: React.FC<{ msg: Message }> = ({ msg }) => {
  const { dispatch } = useContext<ContextType>(ContextApp)
  
  if (!msg) return null
  
  const handleClick = ():void => {
    if (dispatch) {
      dispatch({type: 'deleteSingle', payload: msg.message})
    } else {
      console.log('Error: Context empty')
    }
  }

  return (
    <StyledCard data-testid='card' type={msg.priority} >
      <CardContent>
        <Typography color="textSecondary" gutterBottom>
          {msg.message}
        </Typography>
      </CardContent>
      <CustomCardActions>
        <CustomButton onClick={handleClick}>Clear</CustomButton>
      </CustomCardActions>
    </StyledCard>
  )
}

export default Card
