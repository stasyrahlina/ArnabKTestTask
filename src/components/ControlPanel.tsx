import React, { useContext, useState } from 'react';
import { unsubscribeFunk } from '../Api';
import { ContextApp, ContextType } from '../context';

import { 
  ButtonContainer,
  CustomButton,
} from '../styles';

const ControlPanel: React.FC<{
  subscribeFunc: ()=>void
}> = ({subscribeFunc}) => {
  const [subscribed, setSubscribed] = useState<boolean>(true)
  const { dispatch } = useContext<ContextType>(ContextApp)

  const switchSubscription = ():void => {
    if (subscribed) {
      unsubscribeFunk();
    } else {
      subscribeFunc();
    }

    setSubscribed(!subscribed);
  };
  
  const handleClick = ():void => {
    if (dispatch) {
      dispatch({type: 'deleteAll'})
    } else {
      console.log('Error: Context empty')
    }
  }

  return (
    <ButtonContainer>
      <CustomButton data-testid='stop-button' primary onClick={() => switchSubscription()}>{subscribed ? 'Stop' : 'Start'}</CustomButton>
      <CustomButton data-testid='clear-button' primary onClick={handleClick}>Clear</CustomButton>
    </ButtonContainer>
  )
}

export default ControlPanel
