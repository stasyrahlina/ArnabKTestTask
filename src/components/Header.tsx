import React from 'react';

import {
  AppTitle,
  StyledHeader,
  FullScreenDivider
} from '../styles';

const Header: React.FC<{}> = () => {
  return  (
    <StyledHeader>
      <AppTitle data-testid='app-title'>Coding Challenge</AppTitle>
      <FullScreenDivider />
    </StyledHeader>
  )
}

export default Header
