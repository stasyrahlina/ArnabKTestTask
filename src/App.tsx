import React, { useReducer, useState, useEffect } from 'react';
import { Message, subscribeFunk } from './Api';
import { ContextApp, initialState, messagesReducer } from './context';

import Popup from './components/Popup';
import Header from './components/Header';
import ControlPanel from './components/ControlPanel';
import MessagesGrid from './components/MessagesGrid';

import './reset.css';
import { AppContainer } from './styles';

const App: React.FC<{}> = () => {
  const [showPopup, setShowPopup] = useState<boolean>(false)
  const [state, dispatch] = useReducer(messagesReducer, initialState);

  const subscribeToUpdates = ():void => {
    subscribeFunk((message: Message) => {
      dispatch({type: 'update', payload: message});
      
      if (message.priority === 0) {
        setShowPopup(true)
      }
    })
  };
  
  useEffect(() => {
    subscribeToUpdates()
    return subscribeToUpdates;
  }, []);

  return (
    <ContextApp.Provider value={{dispatch, state}}>
      <AppContainer>
        <Popup open={showPopup} onClose={() => setShowPopup(false)}/>
        <Header />
        <ControlPanel subscribeFunc={subscribeToUpdates} />
        <MessagesGrid />
      </AppContainer>
    </ContextApp.Provider>
  );
}

export default App;
