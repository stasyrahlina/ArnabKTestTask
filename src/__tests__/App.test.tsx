import React from 'react';
import '@testing-library/jest-dom'
import {render, waitFor} from '@testing-library/react'
import renderer from 'react-test-renderer';

import App from '../App';
import Card from '../components/Card';
import Header from '../components/Header';
import ControlPanel from '../components/ControlPanel';

test('renders learn react link', () => {
  const comp = render(<App />);
  expect(comp).toBeTruthy();
});

test('Header rendered header text', () => {
  const {getByTestId} = render(<Header />);
    expect(getByTestId('app-title')).toBeInTheDocument();
});

test('ControlPanel rendered buttons', () => {
  const {getByTestId} = render(<ControlPanel subscribeFunc={()=>{}} />);
    expect(getByTestId('stop-button')).toBeInTheDocument();
    expect(getByTestId('clear-button')).toBeInTheDocument();
});

test('Messages are rendered for error type', async () => {
  const {queryByTestId } = render(<App />);
    await waitFor(() => {
      expect(queryByTestId('card')).not.toEqual(null);
    })
})

it('Card snapshot', () => {
  const tree = renderer
    .create(<Card msg={
      {
        message: 'some string',
        priority: 0
      }
    } />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Header snapshot', () => {
  const tree = renderer
    .create(<Header />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('ControlPanel snapshot', () => {
  const tree = renderer
    .create(<ControlPanel subscribeFunc={()=>{}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
