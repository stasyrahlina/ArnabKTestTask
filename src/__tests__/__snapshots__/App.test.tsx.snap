// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1937z50-MuiPaper-root-MuiCard-root"
  data-testid="card"
  type={0}
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-1ydyjjr-MuiTypography-root"
    >
      some string
    </p>
  </div>
  <div
    className="MuiCardActions-root MuiCardActions-spacing css-la0oad-MuiCardActions-root"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-fmolty-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Clear
    </button>
  </div>
</div>
`;

exports[`ControlPanel snapshot 1`] = `
<div
  className="css-edgrgi"
>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-hhgmux-MuiButtonBase-root-MuiButton-root"
    data-testid="stop-button"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Stop
  </button>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-hhgmux-MuiButtonBase-root-MuiButton-root"
    data-testid="clear-button"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Header snapshot 1`] = `
<header
  className="css-1e2fy0l"
>
  <h1
    className="css-5od3bw"
    data-testid="app-title"
  >
    Coding Challenge
  </h1>
  <hr
    className="MuiDivider-root MuiDivider-fullWidth css-iovw5x-MuiDivider-root"
  />
</header>
`;
